autoblocks/_impl/testing/run.py

```diff
@@ -0,0 +1,285 @@
+    await asyncio.gather(*(sem_coro(c) for c in coroutines), return_exceptions=True)
+
+
+async def evaluate_output(
+    test_id: str,
+    test_case: BaseTestCase,
+    output: Any,
+    evaluator: BaseEvaluator,
+):
+    evaluation: Optional[Evaluation] = None
+
+    if inspect.iscoroutinefunction(evaluator.evaluate):
+        try:
+            evaluation = await evaluator.evaluate(test_case, output)
+        except Exception as err:
+            await send_error(
+                test_id=test_id,
+                test_case_hash=test_case._cached_hash,
+                evaluator_id=evaluator.id,
+                error=err,
+            )
+    else:
+        try:
+            ctx = contextvars.copy_context()
+            evaluation = await loop.run_in_executor(
+                None,
+                ctx.run,
+                evaluator.evaluate,
+                test_case,
+                output,
+            )
+        except Exception as err:
+            await send_error(
+                test_id=test_id,
+                test_case_hash=test_case._cached_hash,
+                evaluator_id=evaluator.id,
+                error=err,
+            )
+
+    if evaluation is None:
+        return
+
+    await client.post(
+        "/evals",
+        json=dict(
+            testExternalId=test_id,
+            testCaseHash=test_case._cached_hash,
+            evaluatorExternalId=evaluator.id,
+            score=evaluation.score,
+            threshold=dataclasses.asdict(evaluation.threshold) if evaluation.threshold else None,
+        ),
+    )
+
+
+async def run_test_case(
+    test_id: str,
+    test_case: BaseTestCase,
+    evaluators: List[BaseEvaluator],
+    fn: Callable,
+    max_evaluator_concurrency: int,
+):
+    current_test_id_var.set(test_id)
+    current_test_case_var.set(test_case)
+
+    output = None
+
+    if inspect.iscoroutinefunction(fn):
+        try:
+            output = await fn(test_case)
+        except Exception as err:
+            await send_error(
+                test_id=test_id,
+                test_case_hash=test_case._cached_hash,
+                evaluator_id=None,
+                error=err,
+            )
+    else:
+        try:
+            ctx = contextvars.copy_context()
+            output = await loop.run_in_executor(None, ctx.run, fn, test_case)
+        except Exception as err:
+            await send_error(
+                test_id=test_id,
+                test_case_hash=test_case._cached_hash,
+                evaluator_id=None,
+                error=err,
+            )
+
+    if output is None:
+        return
+
+    await client.post(
+        "/results",
+        json=dict(
+            testExternalId=test_id,
+            testCaseHash=test_case._cached_hash,
+            testCaseBody=dataclasses.asdict(test_case),
+            testCaseOutput=output,
+        ),
+    )
+
+    await gather_with_max_concurrency(
```

Comment:
do we need to handle an exception being returned here and send it to `/errors`?

Reply:
that's handled in the function it's calling: 

```python
        try:
            output = await fn(test_case)
        except Exception as err:
            await send_error(
                test_id=test_id,
                test_case_hash=test_case._cached_hash,
                evaluator_id=None,
                error=err,
            )
```

and we shouldn't need to try/catch `gather_with_max_concurrency` b/c we're setting return_exceptions=True, which makes it like `Promise.allSettled`. but now that you mention it, maybe a cleaner way to do this would be to let the called function throw and then handle the exception when it's returned from `gather_with_max_concurrency`

Reply:
hmm might punt on this, looks like the stacktrace isn't propagated and gather() waits for everything to complete, would need to figure out how to propagate the stacktrace and find something that yields as things complete

Reply:
Ya I was mainly wondering if we needed to handle asyncio throwing an error in that function

Reply:
ok i can add that but won't be able to add a test i don't think since we're catching all of the errors that we can mock in a test